app/custom_domain_utils.py

```diff
@@ -0,0 +1,105 @@
+import re
+
+from dataclasses import dataclass
+from typing import Optional
+
+from app.db import Session
+from app.email_utils import get_email_domain_part
+from app.log import LOG
+from app.models import User, CustomDomain, SLDomain, Mailbox
+
+_ALLOWED_DOMAIN_REGEX = re.compile(r"^(?!-)[A-Za-z0-9-]{1,63}(?<!-)$")
+
+
+@dataclass
+class CreateCustomDomainResult:
+    message: str = ""
+    message_category: str = ""
+    success: bool = False
+    instance: Optional[CustomDomain] = None
+    redirect: Optional[str] = None
+
+
+def is_valid_domain(domain: str) -> bool:
+    """
+    Checks that a domain is valid according to RFC 1035
+    """
+    if len(domain) > 255:
+        return False
+    if domain.endswith("."):
+        domain = domain[:-1]  # Strip the trailing dot
+    labels = domain.split(".")
+    if not labels:
+        return False
+    for label in labels:
+        if not _ALLOWED_DOMAIN_REGEX.match(label):
+            return False
+    return True
+
+
+def sanitize_domain(domain: str) -> str:
+    new_domain = domain.lower().strip()
+    if new_domain.startswith("http://"):
+        new_domain = new_domain[len("http://") :]
+
+    if new_domain.startswith("https://"):
+        new_domain = new_domain[len("https://") :]
+
+    return new_domain
+
+
+def can_domain_be_used(user: User, domain: str) -> Optional[str]:
+    if not is_valid_domain(domain):
+        return "This is not a valid domain"
+    elif SLDomain.get_by(domain=domain):
+        return "A custom domain cannot be a built-in domain."
+    elif CustomDomain.get_by(domain=domain):
+        return f"{domain} already used"
+    elif get_email_domain_part(user.email) == domain:
+        return "You cannot add a domain that you are currently using for your personal email. Please change your personal email to your real email"
+    elif Mailbox.filter(
+        Mailbox.verified.is_(True), Mailbox.email.endswith(f"@{domain}")
```

Comment:
Note: We may need to check if postgres allows to have two indexes over a field to have a `gin` index on the mailbox email so that we can do efficient `LIKE` queries.